File: lib/features/player/models/war_stats_filter.dart
class WarStatsFilter {
  final String? season;
  final DateTime? startDate;
  final DateTime? endDate;
  final int? ownTownHall;
  final int? enemyTownHall;
  final List<int>? ownTownHalls;
  final List<int>? enemyTownHalls;
  final bool sameTownHall;
  final String warType;
  final List<String>? warTypes;
  final bool? freshAttacksOnly;
  final int? minStars;
  final int? maxStars;
  final List<int>? allowedStars;
  final double? minDestruction;
  final double? maxDestruction;
  final int? minMapPosition;
  final int? maxMapPosition;
  final int limit;
  final Map<String, dynamic>? metadata;

  const WarStatsFilter({
    this.season,
    this.startDate,
    this.endDate,
    this.ownTownHall,
    this.enemyTownHall,
    this.ownTownHalls,
    this.enemyTownHalls,
    this.sameTownHall = false,
    this.warType = "all",
    this.warTypes,
    this.freshAttacksOnly,
    this.minStars,
    this.maxStars,
    this.allowedStars,
    this.minDestruction,
    this.maxDestruction,
    this.minMapPosition,
    this.maxMapPosition,
    this.limit = 50,
    this.metadata,
  });

  factory WarStatsFilter.fromJson(Map<String, dynamic> json) {
    return WarStatsFilter(
      season: json['season'],
      startDate: json['timestamp_start'] != null 
          ? DateTime.fromMillisecondsSinceEpoch(json['timestamp_start'] * 1000)
          : null,
      endDate: json['timestamp_end'] != null 
          ? DateTime.fromMillisecondsSinceEpoch(json['timestamp_end'] * 1000)
          : null,
      ownTownHall: json['own_th'] is List && (json['own_th'] as List).length == 1
          ? (json['own_th'] as List)[0]
          : null,
      enemyTownHall: json['enemy_th'] is List && (json['enemy_th'] as List).length == 1
          ? (json['enemy_th'] as List)[0]
          : null,
      ownTownHalls: json['own_th'] is List ? List<int>.from(json['own_th']) : null,
      enemyTownHalls: json['enemy_th'] is List ? List<int>.from(json['enemy_th']) : null,
      sameTownHall: json['same_th'] ?? false,
      warType: json['type'] is List && (json['type'] as List).length == 1
          ? (json['type'] as List)[0]
          : "all",
      warTypes: json['type'] is List ? List<String>.from(json['type']) : null,
      freshAttacksOnly: json['fresh_only'],
      allowedStars: json['stars'] is List ? List<int>.from(json['stars']) : null,
      minDestruction: json['min_destruction']?.toDouble(),
      maxDestruction: json['max_destruction']?.toDouble(),
      minMapPosition: json['map_position_min'],
      maxMapPosition: json['map_position_max'],
      limit: json['limit'] ?? 50,
      metadata: json['metadata'] != null ? Map<String, dynamic>.from(json['metadata']) : null,
    );
  }

  Map<String, dynamic> toJson() {
    final Map<String, dynamic> data = {
      'limit': limit,
      'same_th': sameTownHall,
    };
    
    if (warTypes != null && warTypes!.isNotEmpty && !warTypes!.contains('all')) {
      data['type'] = warTypes;
    } else if (warType != "all") {
      data['type'] = warType;
    }

    if (season != null) {
      data['season'] = season;
    }
    if (startDate != null) {
      data['timestamp_start'] = startDate!.millisecondsSinceEpoch ~/ 1000;
    }
    if (endDate != null) {
      data['timestamp_end'] = endDate!.millisecondsSinceEpoch ~/ 1000;
    }
    if (ownTownHalls != null && ownTownHalls!.isNotEmpty) {
      data['own_th'] = ownTownHalls;
    } else if (ownTownHall != null) {
      data['own_th'] = ownTownHall;
    }
    if (enemyTownHalls != null && enemyTownHalls!.isNotEmpty) {
      data['enemy_th'] = enemyTownHalls;
    } else if (enemyTownHall != null) {
      data['enemy_th'] = enemyTownHall;
    }
    if (freshAttacksOnly != null) {
      data['fresh_only'] = freshAttacksOnly;
    }
    if (allowedStars != null && allowedStars!.isNotEmpty) {
      data['stars'] = allowedStars;
    } else {
      if (minStars != null) {
        data['min_stars'] = minStars;
      }
      if (maxStars != null) {
        data['max_stars'] = maxStars;
      }
    }
    if (minDestruction != null) {
      data['min_destruction'] = minDestruction;
    }
    if (maxDestruction != null) {
      data['max_destruction'] = maxDestruction;
    }
    if (minMapPosition != null) {
      data['map_position_min'] = minMapPosition;
    }
    if (maxMapPosition != null) {
      data['map_position_max'] = maxMapPosition;
    }
    if (metadata != null) {
      data['metadata'] = metadata;
    }

    return data;
  }

  WarStatsFilter copyWith({
    String? season,
    DateTime? startDate,
    DateTime? endDate,
    int? ownTownHall,
    int? enemyTownHall,
    List<int>? ownTownHalls,
    List<int>? enemyTownHalls,
    bool? sameTownHall,
    String? warType,
    List<String>? warTypes,
    bool? freshAttacksOnly,
    int? minStars,
    int? maxStars,
    List<int>? allowedStars,
    double? minDestruction,
    double? maxDestruction,
    int? minMapPosition,
    int? maxMapPosition,
    int? limit,
    Map<String, dynamic>? metadata,
  }) {
    return WarStatsFilter(
      season: season ?? this.season,
      startDate: startDate ?? this.startDate,
      endDate: endDate ?? this.endDate,
      ownTownHall: ownTownHall ?? this.ownTownHall,
      enemyTownHall: enemyTownHall ?? this.enemyTownHall,
      ownTownHalls: ownTownHalls ?? this.ownTownHalls,
      enemyTownHalls: enemyTownHalls ?? this.enemyTownHalls,
      sameTownHall: sameTownHall ?? this.sameTownHall,
      warType: warType ?? this.warType,
      warTypes: warTypes ?? this.warTypes,
      freshAttacksOnly: freshAttacksOnly ?? this.freshAttacksOnly,
      minStars: minStars ?? this.minStars,
      maxStars: maxStars ?? this.maxStars,
      allowedStars: allowedStars ?? this.allowedStars,
      minDestruction: minDestruction ?? this.minDestruction,
      maxDestruction: maxDestruction ?? this.maxDestruction,
      minMapPosition: minMapPosition ?? this.minMapPosition,
      maxMapPosition: maxMapPosition ?? this.maxMapPosition,
      limit: limit ?? this.limit,
      metadata: metadata ?? this.metadata,
    );
  }

  /// Default filter for mobile app
  static WarStatsFilter defaultFilter() {
    return WarStatsFilter(
      startDate: DateTime.now().subtract(const Duration(days: 180)), // 6 months ago
      endDate: DateTime.now(),
      warType: "all",
      limit: 50,
    );
  }

  /// Check if any filters are active (not default)
  bool hasActiveFilters() {
    return season != null ||
        startDate != null ||
        endDate != null ||
        ownTownHall != null ||
        enemyTownHall != null ||
        (ownTownHalls != null && ownTownHalls!.isNotEmpty) ||
        (enemyTownHalls != null && enemyTownHalls!.isNotEmpty) ||
        sameTownHall ||
        warType != "all" ||
        (warTypes != null && warTypes!.isNotEmpty) ||
        freshAttacksOnly != null ||
        minStars != null ||
        maxStars != null ||
        (allowedStars != null && allowedStars!.isNotEmpty) ||
        minDestruction != null ||
        maxDestruction != null ||
        minMapPosition != null ||
        maxMapPosition != null;
  }

  /// Get filter summary text
  String getFilterSummary() {
    List<String> filters = [];
    
    if (season != null) {
      filters.add("Season $season");
    }
    
    // Add date range filter display
    if (startDate != null && endDate != null) {
      final start = "${startDate!.day}/${startDate!.month}/${startDate!.year}";
      final end = "${endDate!.day}/${endDate!.month}/${endDate!.year}";
      filters.add("$start - $end");
    } else if (startDate != null) {
      final start = "${startDate!.day}/${startDate!.month}/${startDate!.year}";
      filters.add("From $start");
    } else if (endDate != null) {
      final end = "${endDate!.day}/${endDate!.month}/${endDate!.year}";
      filters.add("Until $end");
    }
    
    if (ownTownHalls != null && ownTownHalls!.isNotEmpty) {
      if (ownTownHalls!.length == 1) {
        filters.add("TH${ownTownHalls!.first} attacks");
      } else {
        filters.add("TH${ownTownHalls!.join(', ')} attacks");
      }
    } else if (ownTownHall != null) {
      filters.add("TH$ownTownHall attacks");
    }
    
    if (enemyTownHalls != null && enemyTownHalls!.isNotEmpty) {
      if (enemyTownHalls!.length == 1) {
        filters.add("vs TH${enemyTownHalls!.first}");
      } else {
        filters.add("vs TH${enemyTownHalls!.join(', ')}");
      }
    } else if (enemyTownHall != null) {
      filters.add("vs TH$enemyTownHall");
    }
    
    if (sameTownHall) filters.add("Same TH only");
    
    if (warTypes != null && warTypes!.isNotEmpty) {
      if (warTypes!.length == 1) {
        filters.add("${warTypes!.first.toUpperCase()} wars");
      } else {
        filters.add("${warTypes!.map((w) => w.toUpperCase()).join(', ')} wars");
      }
    } else if (warType != "all") {
      filters.add("${warType.toUpperCase()} wars");
    }
    
    if (freshAttacksOnly == true) filters.add("Fresh attacks only");
    
    if (allowedStars != null && allowedStars!.isNotEmpty) {
      if (allowedStars!.length == 1) {
        filters.add("${allowedStars!.first} ⭐ only");
      } else {
        filters.add("${allowedStars!.join(', ')} ⭐ only");
      }
    } else if (minStars != null || maxStars != null) {
      if (minStars != null && maxStars != null) {
        filters.add("$minStars-$maxStars stars");
      } else if (minStars != null) {
        filters.add("$minStars+ stars");
      } else {
        filters.add("≤$maxStars stars");
      }
    }
    if (minDestruction != null || maxDestruction != null) {
      if (minDestruction != null && maxDestruction != null) {
        filters.add("${minDestruction!.toStringAsFixed(0)}-${maxDestruction!.toStringAsFixed(0)}% destruction");
      } else if (minDestruction != null) {
        filters.add("${minDestruction!.toStringAsFixed(0)}%+ destruction");
      } else {
        filters.add("≤${maxDestruction!.toStringAsFixed(0)}% destruction");
      }
    }
    
    return filters.isEmpty ? "No filters applied" : filters.join(", ");
  }
}

File: expo/src/features/player/models/war-stats-filter.ts
import { bool, int, isRecord, nullableInt, nullableString, type JsonRecord } from './parsing';
export interface WarStatsFilterInput {
  season?: string | null;
  startDate?: Date | null;
  endDate?: Date | null;
  ownTownHall?: number | null;
  enemyTownHall?: number | null;
  ownTownHalls?: readonly number[] | null;
  enemyTownHalls?: readonly number[] | null;
  sameTownHall?: boolean;
  warType?: string;
  warTypes?: readonly string[] | null;
  freshAttacksOnly?: boolean | null;
  minStars?: number | null;
  maxStars?: number | null;
  allowedStars?: readonly number[] | null;
  minDestruction?: number | null;
  maxDestruction?: number | null;
  minMapPosition?: number | null;
  maxMapPosition?: number | null;
  limit?: number;
  metadata?: JsonRecord | null;
}
export class WarStatsFilter {
  readonly season: string | null;
  readonly startDate: Date | null;
  readonly endDate: Date | null;
  readonly ownTownHall: number | null;
  readonly enemyTownHall: number | null;
  readonly ownTownHalls: readonly number[] | null;
  readonly enemyTownHalls: readonly number[] | null;
  readonly sameTownHall: boolean;
  readonly warType: string;
  readonly warTypes: readonly string[] | null;
  readonly freshAttacksOnly: boolean | null;
  readonly minStars: number | null;
  readonly maxStars: number | null;
  readonly allowedStars: readonly number[] | null;
  readonly minDestruction: number | null;
  readonly maxDestruction: number | null;
  readonly minMapPosition: number | null;
  readonly maxMapPosition: number | null;
  readonly limit: number;
  readonly metadata: JsonRecord | null;
  constructor(i: WarStatsFilterInput = {}) {
    this.season = i.season ?? null;
    this.startDate = i.startDate ?? null;
    this.endDate = i.endDate ?? null;
    this.ownTownHall = i.ownTownHall ?? null;
    this.enemyTownHall = i.enemyTownHall ?? null;
    this.ownTownHalls = i.ownTownHalls ?? null;
    this.enemyTownHalls = i.enemyTownHalls ?? null;
    this.sameTownHall = i.sameTownHall ?? false;
    this.warType = i.warType ?? 'all';
    this.warTypes = i.warTypes ?? null;
    this.freshAttacksOnly = i.freshAttacksOnly ?? null;
    this.minStars = i.minStars ?? null;
    this.maxStars = i.maxStars ?? null;
    this.allowedStars = i.allowedStars ?? null;
    this.minDestruction = i.minDestruction ?? null;
    this.maxDestruction = i.maxDestruction ?? null;
    this.minMapPosition = i.minMapPosition ?? null;
    this.maxMapPosition = i.maxMapPosition ?? null;
    this.limit = i.limit ?? 1000;
    this.metadata = i.metadata ?? null;
  }
  static fromJson(j: JsonRecord) {
    const own = Array.isArray(j.own_th) ? j.own_th.map((value) => int(value)) : null,
      enemy = Array.isArray(j.enemy_th) ? j.enemy_th.map((value) => int(value)) : null,
      types = Array.isArray(j.type) ? j.type.map(String) : null;
    return new WarStatsFilter({
      season: nullableString(j.season),
      startDate: typeof j.timestamp_start === 'number' ? new Date(j.timestamp_start * 1000) : null,
      endDate: typeof j.timestamp_end === 'number' ? new Date(j.timestamp_end * 1000) : null,
      ownTownHall: own?.length === 1 ? own[0]! : null,
      enemyTownHall: enemy?.length === 1 ? enemy[0]! : null,
      ownTownHalls: own,
      enemyTownHalls: enemy,
      sameTownHall: bool(j.same_th),
      warType: types?.length === 1 ? types[0]! : 'all',
      warTypes: types,
      freshAttacksOnly: typeof j.fresh_only === 'boolean' ? j.fresh_only : null,
      allowedStars: Array.isArray(j.stars) ? j.stars.map((value) => int(value)) : null,
      minDestruction: typeof j.min_destruction === 'number' ? j.min_destruction : null,
      maxDestruction: typeof j.max_destruction === 'number' ? j.max_destruction : null,
      minMapPosition: nullableInt(j.map_position_min),
      maxMapPosition: nullableInt(j.map_position_max),
      limit: int(j.limit, 1000),
      metadata: isRecord(j.metadata) ? j.metadata : null,
    });
  }
  toJson(): JsonRecord {
    const d: JsonRecord = { same_th: this.sameTownHall, limit: this.limit };
    if (this.warTypes?.length && !this.warTypes.includes('all')) d.type = this.warTypes;
    else if (this.warType !== 'all') d.type = this.warType;
    if (this.season) d.season = this.season;
    if (this.startDate) d.timestamp_start = Math.trunc(this.startDate.getTime() / 1000);
    if (this.endDate) d.timestamp_end = Math.trunc(this.endDate.getTime() / 1000);
    if (this.ownTownHalls?.length) d.own_th = this.ownTownHalls;
    else if (this.ownTownHall !== null) d.own_th = this.ownTownHall;
    if (this.enemyTownHalls?.length) d.enemy_th = this.enemyTownHalls;
    else if (this.enemyTownHall !== null) d.enemy_th = this.enemyTownHall;
    if (this.freshAttacksOnly !== null) d.fresh_only = this.freshAttacksOnly;
    if (this.allowedStars?.length) d.stars = this.allowedStars;
    else {
      if (this.minStars !== null) d.min_stars = this.minStars;
      if (this.maxStars !== null) d.max_stars = this.maxStars;
    }
    if (this.minDestruction !== null) d.min_destruction = this.minDestruction;
    if (this.maxDestruction !== null) d.max_destruction = this.maxDestruction;
    if (this.minMapPosition !== null) d.map_position_min = this.minMapPosition;
    if (this.maxMapPosition !== null) d.map_position_max = this.maxMapPosition;
    if (this.metadata) d.metadata = this.metadata;
    return d;
  }
  static defaultFilter() {
    return new WarStatsFilter({ warType: 'all', limit: 50 });
  }
  hasActiveFilters() {
    return (
      this.season !== null ||
      this.startDate !== null ||
      this.endDate !== null ||
      this.ownTownHall !== null ||
      this.enemyTownHall !== null ||
      !!this.ownTownHalls?.length ||
      !!this.enemyTownHalls?.length ||
      this.sameTownHall ||
      this.warType !== 'all' ||
      !!this.warTypes?.length ||
      this.freshAttacksOnly !== null ||
      this.minStars !== null ||
      this.maxStars !== null ||
      !!this.allowedStars?.length ||
      this.minDestruction !== null ||
      this.maxDestruction !== null ||
      this.minMapPosition !== null ||
      this.maxMapPosition !== null
    );
  }
}
